clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -580,20 +595,33 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
         return new ClusterAndWaitTime(cluster, elapsed);
     }
 
+    private long waitOnPid(long maxWaitMs) throws InterruptedException {
+        if (transactionState == null) {
```

Comment:
I'll change this to `maybeWaitOnPid`. Would prefer to isolate all the logic into this function, since it's called multiple times.